File: packages/xbridge-android/src/main/java/io/xbridge/XBridgeNativeBridge.kt
package io.xbridge

/**
 * Delegate interface that the host app implements to forward XBridge method
 * calls to its existing native bridge (e.g. a DsBridge plugin instance).
 *
 * This is the **seam** that keeps XBridge business-free: the SDK itself has
 * zero knowledge of any business method. The app
 * supplies a single implementation that routes every `method` string to the
 * appropriate legacy handler.
 *
 * ## Threading
 *
 * [invoke] may be called from:
 * - The Flutter platform thread (MethodChannel callback thread).
 * - The WebView's JS thread (via [XBridgeSyncInterface.callSync]).
 *
 * Implementations must be thread-safe or must dispatch to their own preferred
 * thread internally. If the legacy bridge requires the UI thread, the
 * implementation should handle that dispatch itself.
 */
interface XBridgeNativeBridge {

    /**
     * Synchronously invoke the native bridge handler for [method].
     *
     * @param method  The business method name (e.g. `"getDeviceInfo"`).
     * @param params  The raw parameters — may be `null`, a primitive, a
     *                `Map<String, Any?>`, or a `List<Any?>` depending on what
     *                the H5 side sent. Implementations interpret this.
     * @return        The result value, or `null` if the method has no return.
     */
    fun invoke(method: String, params: Any?): Any?
}

File: packages/xbridge-ios/Sources/XBridgeiOS/XBridgeNativeBridge.swift
// SPDX-License-Identifier: MIT
//
// XBridgeNativeBridge.swift
// XBridgeiOS
//
// Created by XBridge SDK on 2024-01-01.
//

import Foundation

/// A generic delegate protocol that the host app implements to forward
/// XBridge fallback calls to its existing bridge handler (e.g. your existing native bridge).
///
/// XBridge itself contains zero business logic. The app supplies an
/// implementation of this protocol that knows how to route `method` strings
/// to its existing native bridge plugin instance.
///
/// Usage:
/// ```swift
/// class MyBridgeAdapter: XBridgeNativeBridge {
///     func invoke(method: String, params: Any?) -> Any? {
///         // Forward to your existing bridge plugin handler
///         return existingBridge.call(method, args: params)
///     }
/// }
/// ```
public protocol XBridgeNativeBridge: AnyObject {
    /// Invoke a bridge method synchronously.
    ///
    /// - Parameters:
    ///   - method: The method name (e.g. `"getDeviceInfo"`, `"getAppInfo"`).
    ///   - params: Arbitrary parameters from the caller. May be nil.
    /// - Returns: The result value, or nil if the method has no return value.
    ///
    /// - Note: Implementations should be thread-safe. If the underlying
    ///   bridge handler requires the main thread, the implementation is
    ///   responsible for dispatching.
    func invoke(method: String, params: Any?) -> Any?
}
